### ISSUE TITLE
Only 50% of the reads are mapped

### ISSUE THREAD

**Issue by harisankarsadasivan** [haowenz/sigmap#11]
i ran with 4 threads and also 16 threads on a 16 core machine, my input is 20 fast5 files of 4K reads each. I only see 20K lines in the paf file, out of which only ~50% are mapped  when guppy+minimap2 would map > 90% of all the reads for the same set. Any suggestions?

**Comment by haowenz** [comment]
I guess more information is needed. What command line did you use? What dataset were you using? How large is the target genome? What is the sequencing chemistry and did you use the corresponds pore model?

Also did you tried Uncalled and see if it works?

**Comment by harisankarsadasivan** [comment]
@haowenz 
Thanks for your response.
I used the commands from the git page (built index first followed by mapping) but with SARS-CoV-2 genome(~30Kbp) on fast5's from r9.4.1 pore. I used [r9.4_180mv_450bps_6mer], I'm thinking it's the same.

Map command I used is:
./sigmap -m -r covid_ref.fasta -p extern/kmer_models/r9.4_180mv_450bps_6mer/template_median68pA.model -x covid_index -s fast5/ -o covid_mapping.paf -t 4

**Comment by haowenz** [comment]
Is this direct RNA sequencing of SARS-CoV-2?

**Comment by harisankarsadasivan** [comment]
@haowenz No, this is rt-cDNA. UNCALLED can map this using same pore model. 
I think I figured the issue with the SIGMAP software: It seems to map all reads in the directory in single thread mode but does not do that with 4 or 16 threads. There's probably some issue with the software with respect to multi-threading.
But still seems to map only ~50% of the total reads from the PAF file in single threaded mode.

**Comment by haowenz** [comment]
> @haowenz No, this is rt-cDNA. UNCALLED can map this using same pore model.

You mean Uncalled can map most of the reads like >90%?

> I think I figured the issue with the SIGMAP software: It seems to map all reads in the directory in single thread mode but does not do that with 4 or 16 threads. There's probably some issue with the software with respect to multi-threading.

Is this for mapping or fast5 loading? The loading is single thread, which is slow. The mapping should be multi-threading unless there is something wrong with your openmp on your system.

> But still seems to map only ~50% of the total reads from the PAF file.

Is this data publicly available? If not, is there a way for you to share a simple? I can take a quick look. Or you can try to tune the parameters a little bit. For example, try a lower "--step-size 1". Or try a larger "--search-radius 0.1".

Btw, are there many repeats in the genome? And you were saying "guppy+minimap2 would map > 90%". Is this Readfish running for selective sequencing or you were using the full length of the reads?

**Comment by harisankarsadasivan** [comment]
@haowenz Please use our publicly available dataset for SARS-CoV-2 on aws from https://github.com/TimD1/SquiggleFilter/blob/master/setup.sh (I recommend following the install instructions on the repo for aws and running the dataset download commands in the setup.sh for the covid dataset). You will find both fast5 and covid reference genome here (average read lengths of covid is ~400bp and human is ~800bp). IIRC UNCALLED maps I think 70-80% using just 2000 samples on this dataset but maps close to 90% with full length raw signal. Guppy + minimap2 on trimmed 200bp (I did not use readfish) maps > 99% of the basecalled reads. I dont think there are many repeats in covid genome.

I checked SIGMAP (without any fine tuning) again and seems to map only ~50% of full length reads.

Additionally, after we solve this, would you be kind enough to briefly explain how I could test SIGMAP on maybe like 2000 samples instead of full length reads?

**Comment by haowenz** [comment]
I did a run and found that lots of mappings are short and just not output by sigmap. I am not sure those mapping are correct. Do you have the fastq file for all these raw signal in fast5? I checked your repo but didn't find it. If you can provide the fastq reads, I will be able verify the mapping and get back to you soon with the reasons.

**Comment by harisankarsadasivan** [comment]
@haowenz  Thanks for looking into it.
I have just mailed it to you on hwzhang@gatech.edu. I'm interested to know how much can SIGMAP be tuned to map these reads of average length ~430bp. 
Additionally, once you get the best tuning params for this dataset, please also let me how to test SIGMAP on a fixed raw signal size of my choice instead of full length of raw signal.

**Comment by haowenz** [comment]
Thank you for sharing. There are only 2893 reads in the fastq file you shared. However, there seems to be 1382016 reads in fast5 files. Can you tell which fast5 file you ran base calling and generated these reads in fastq file?

**Comment by harisankarsadasivan** [comment]
@haowenz I just re-downloaded the file I shared and checked, there are 5317216 lines or 1.32M reads.  This is the entire basecall.
The gzip file on download is 674MB. Kindly check at your end.

**Comment by haowenz** [comment]
My downloaded file was broken. I downloaded it again and found it complete. I tried to map the whole dataset using the default parameters and as you said only around 50% reads were in the mapping output. It took 3.6h to load the fast5 files and only 100s to map all the reads. And I saw many read lengths were very short and for this reason even they were mapped, they were not output by Sigmap.

Since the loading time is too long, I have to test on a sample with 1,0000 reads. And I changed the parameters (adding `--min-num-anchors-output 2`) to output more short mappings. The mapping rate for this sample increased from 47.37% to 82.29%. Moreover, since the read length is small, I also reduce the step size to 1 (`--step-size 1`) to get enough seeds. And the mapping rate further increased to 85.17%. By checking with minimap2 mapping output, I was also able to confirm that most of sigmap mapping output is correct. 

There are more parameters that are hard-coded in the program and you cannot tune them simply with parameters in command lines. But I tried to change one of them and the change easily lead to >90% mapping rate. However, those output reads might be in repeats and their mappings might not be reliable. That's why in the code they were filtered out for the purpose of selective sequencing in practice.

In short, you can try to run sigmap with additional parameters setting `--min-num-anchors-output 2 --step-size 1` for small genome and shorter reads on your whole dataset. I may later add these to the document.

**Comment by harisankarsadasivan** [comment]
Thanks @haowenz . Marking this issue as closed.

**Comment by haowenz** [comment]
> Additionally, once you get the best tuning params for this dataset, please also let me how to test SIGMAP on a fixed raw signal size of my choice instead of full length of raw signal.

I am not sure if I get your question correctly. What Sigmap does is that it maps the signal chunk by chunk, and once the best mapping is much better than the second best mapping, it will stop and output the mapping. If you want to control the number of chunks that you want to use in the mapping given you already have the reads in full length, you can set --stop-mapping and --stop-mapping-mean to a large number, and --min-num-anchors and together with --max-num-chunks to the same very large number too. In this case, the mapping may not stop early so that the number of specified chunks might be used in the mapping. Since Sigmap is not designed for this purpose, I never tried this and not sure if it would work. But you may give it a try if you want.

**Comment by harisankarsadasivan** [comment]
@haowenz I get the error:
```sigmap: src/event.h:188: sigmap::Event sigmap::CreateEvent(size_t, size_t, const float*, const float*, size_t): Assertion `start < signal_length' failed.
When i change
